### ISSUE TITLE
Numpy memmap errors

### ISSUE THREAD

**Issue by juliotux** [pympler/pympler#139]
Pymbler is presenting errors when try to profile an object with a [`numpy.memmap`](https://numpy.org/doc/stable/reference/generated/numpy.memmap.html) object. Executing the following command:

```
from pympler import asizeof
import numpy as np

m = np.memmap('./test.npy', mode='w+', dtype='f8', shape=(10000, 10000))
print(asizeof.asizeof(m))
```

Results in the following error:

```
...

File ~/.anaconda/envs/astropop/lib/python3.10/site-packages/pympler/asizeof.py:1597, in _typedef(obj, derive, frames, infer)
   1595     v.set(refs=_namedtuple_refs)
   1596 elif numpy and _isnumpy(obj):  # NumPy data
-> 1597     v.set(**_numpy_kwds(obj))
   1598     if _numpy_excl:
   1599         _getsizeof_excls_add(t)

File ~/.anaconda/envs/astropop/lib/python3.10/site-packages/pympler/asizeof.py:1183, in _Typedef.set(self, safe_len, **kwds)
   1181     d = self.kwds()
   1182     d.update(kwds)
-> 1183     self.reset(**d)
   1184 if safe_len and self.item:
   1185     self.leng = _len

File ~/.anaconda/envs/astropop/lib/python3.10/site-packages/pympler/asizeof.py:1192, in _Typedef.reset(self, base, item, leng, refs, both, kind, type, vari)
   1189 '''Reset all specified attributes.
   1190 '''
   1191 if base < 0:
-> 1192     raise ValueError('invalid option: %s=%r' % ('base', base))
   1193 else:
   1194     self.base = base

ValueError: invalid option: base=-799999856
```

As memmap main usage is for high memory usage optimizing, support it could be a good addition to Pympler.

**Comment by mrJean1** [comment]
Which Pythin version is this and what is the value of `m.nbytes`?

**Comment by juliotux** [comment]
For the exactly example I wrote in the first comment, the `m.nbytes` show:

```
In [3]: m.nbytes
Out[3]: 800000000
```

It is strange to see how `base` is very close to `-m.nbytes`. I repeated the process with a different shape in array and, again, `shape=(100, 100)`, `m.nbytes=80000` and `base=-79856`.

In both cases the difference is the same: 144 bytes. Since `memmap` stores the data in disk (and not in memory), don't the correct size exclude the disk data size? Do the actual memory size to be 144 bytes and the `pympler` tries to subtract the `.nbytes` from that size?

**Comment by mrJean1** [comment]
Very interesting, thank you for following up!  That `base` value for all `numpy` objects is computed in the following function in `asizeof`:
```
    def _numpy_kwds(obj):
        b = _getsizeof(obj, 96) - obj.nbytes  # XXX 96..144 typical?
        ...
```
Normally, `b` is positive for `numpy` objects because `_getsizeof(obj, ...)` -which is just `sys.getsizeof`- _includes_ the value of `obj.nbytes`.  But appearently, it does not for `numpy.memmap` objects.  That may be one of the [unpleasant interactions](https://numpy.org/doc/stable/reference/generated/numpy.memmap.html) of `numpy.memmap` objects.

Now, the question is if and how to handle this?  My initial take was to simply not size `numpy.memmap` objects at all, because `memmap`-ed files are rarely mapped entirely into memory.  If they were, that would defeat the whole purpose of memory-mapped files since that would be no different from just reading the entire file.  There are system calls to get the amount of virtual memory used by memory-mapped files, like this [mincore](https://www.ibm.com/docs/en/aix/7.2?topic=m-mincore-subroutine).

But apart from that, if `asizeof` should size `numpy.memmap` objects, one way to do so could be:
```
    def _numpy_kwds(obj):
        if isinstance(obj, numpy.memmap):
            b = _getsizeof(obj, 96) + int(obj.nbytes * p_memmap)
        else:
            b = _getsizeof(obj, 96) - obj.nbytes  # XXX 96..144 typical?
        ...
```
where `0 <= p_memmap <= 1` is some percentage configurable with an env var or keyword argument.  Then, users could choose the extent to which a `numpy.memmap` is sized in `asizeof`.  Would that be useful?

**Comment by juliotux** [comment]
Yes! This is a very good solution for the problem. At least, it can give the used memory for data fully flushed by setting `p_memmap` to 0.

**Comment by mrJean1** [comment]
Another patch for `asizeof` is forthcoming including a memory-mapped-percentage option.  It will probably a new function to get/set that percentage and the default will be small, perhaps 4%.

But one comment, `asizeof` measures the approximate (virtual) memory used by Python `objects`.  _Not_ the total number of bytes read from or written to files, received from or sent to servers, etc. by the Python application.

**Comment by mrJean1** [comment]
An improved patch has been submitted, merge pending.  Following is a copy of the attached comments.

Sizing `numpy.memmap` objects is handled by taking a _configurable_ percentage of the mapped file size. Mapped files are rarely completely mapped into memory. The memory usage of a `numpy.memmap` object (and any `mmap`-ed file) is typically only a small portion of the file size. That portion depends on the file, the O/S and perhaps other system parameters.

A new `asizeof` function called `amapped` can be use to set that portion as a percentage (and get the previously set percentage). In addition, the size will be rounded up to a multiple of 4Ki bytes, assuming 4 KiB is a reasonable (although O/S dependent) value for the page size of virtual memory.

**Comment by mrJean1** [comment]
The previous patch wasn't quite correct and another patch has been submitted.  Following is that explanation.

Sizing `numpy.memmap` objects is handled by taking a _configurable_ percentage of the `mmap`-ed file size. `Mmap`-ed files are rarely completely loaded into memory.  The memory usage of a `numpy.memmap` object (and any `mmap`-ed file) is typically only a small portion of the file size. That portion depends on the file, the O/S and other system parameters.

A new function called `asizeof.amapped` allows setting that portion as a percentage (and getting the previous percentage). In addition, the size of `numpy.memmap` objects will be rounded up to a multiple of `mmap.PAGESIZE`, the O/S-dependent virtual memory page size.

Sizing other `numpy` objects has been simplified for previously defined `numpy` types.  Plus that should work for any `numpy` type with an instance attribute called `nbytes`.  Lastly, sizing some `numpy` examples can be shown if invoked as follows:
```
python3  asizeof.py  -numpy ...
```
